server/src/test/java/org/elasticsearch/action/admin/cluster/bootstrap/TransportGetDiscoveredNodesActionTests.java

```diff
@@ -0,0 +1,332 @@
+
+            @Override
+            public void start() {
+                throw new AssertionError("should not be called");
+            }
+
+            @Override
+            public void stop() {
+                throw new AssertionError("should not be called");
+            }
+
+            @Override
+            public void close() {
+                throw new AssertionError("should not be called");
+            }
+        };
+        final TransportGetDiscoveredNodesAction transportGetDiscoveredNodesAction
+            = new TransportGetDiscoveredNodesAction(Settings.EMPTY, mock(ActionFilters.class), transportService, discovery);
+
+        final ActionListener<GetDiscoveredNodesResponse> listener = new ActionListener<GetDiscoveredNodesResponse>() {
+            @Override
+            public void onResponse(GetDiscoveredNodesResponse getDiscoveredNodesResponse) {
+                throw new AssertionError("should not be called");
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                throw new AssertionError("should not be called");
+            }
+        };
+
+        assertThat(expectThrows(IllegalStateException.class,
+            () -> transportGetDiscoveredNodesAction.doExecute(mock(Task.class), new GetDiscoveredNodesRequest(), listener))
+            .getMessage(), equalTo("cannot execute a Zen2 action if not using Zen2"));
+
+        threadPool.shutdown();
+    }
+
+    public void testFailsOnNonMasterEligibleNodes() {
+        final DiscoveryNode discoveryNode
+            = new DiscoveryNode("local", buildNewFakeTransportAddress(), emptyMap(), emptySet(), Version.CURRENT);
+
+        final MockTransport transport = new MockTransport();
+        final ThreadPool threadPool = new TestThreadPool("test", Settings.EMPTY);
+        final TransportService transportService = transport.createTransportService(Settings.EMPTY, threadPool,
+            TransportService.NOOP_TRANSPORT_INTERCEPTOR, boundTransportAddress -> discoveryNode, null, emptySet());
+        transportService.start();
+        transportService.acceptIncomingRequests();
+
+        final ClusterSettings clusterSettings = new ClusterSettings(Settings.EMPTY, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS);
+        final ClusterState state = ClusterState.builder(new ClusterName("cluster")).build();
+        final Coordinator coordinator = new Coordinator("local", Settings.EMPTY, clusterSettings, transportService,
+            ESAllocationTestCase.createAllocationService(Settings.EMPTY),
+            new MasterService("local", Settings.EMPTY, threadPool),
+            () -> new InMemoryPersistedState(0, state), r -> emptyList(),
+            new NoOpClusterApplier(), random());
+        coordinator.start();
+
+        final TransportGetDiscoveredNodesAction transportGetDiscoveredNodesAction
+            = new TransportGetDiscoveredNodesAction(Settings.EMPTY, mock(ActionFilters.class), transportService, coordinator);
+
```

Comment:
Same here can we refactor initialization part in a separate method?

Reply:
I did so in fc2f125baa735f4a889a57b959337f3578588f91. I'm not sure I wholly like it, because of things like the fact that we have to change the local node and rely on the transport service not picking it up until started, and also we have to create a transport that responds to handshakes in all cases, but I don't have a strong opinion about this.